### ISSUE TITLE
New command: `spago publish`

### ISSUE THREAD

**Issue by thomashoneyman** [purescript/spago#766]
Related to #204 and https://github.com/purescript/registry/pull/140 and depends on the [new registry](https://github.com/purescript/registry/).

The PureScript registry supports [a minimal API consisting of package additions, updates, and unpublishing](https://github.com/purescript/registry#the-registry-api). This issue tracks adding a command to Spago that at least would support adding and updating packages. To do this, we'll need to support a few things from the registry:

1. The registry uses a new manifest format as described in [the `Manifest.dhall` definition](https://github.com/purescript/registry/blob/master/v1/Manifest.dhall). Spago will need to be able to generate this format from `spago.dhall` files, and therefore `spago.dhall` files themselves probably have to be changed. (This could be tracked in a separate issue if necessary.)
2. The registry adds packages based on [the `Operation.dhall` definition](https://github.com/purescript/registry/blob/master/v1/Operation.dhall). The `spago publish` command would need to produce the correct JSON corresponding with the update or addition operations.
3. The registry relies on issue creation and comments on the registry repository to kick off operations. Spago will need to open an issue on GitHub on the user's behalf, either automatically (with a confirmation) or by providing the user with a link to create the issue with the correct JSON contents themselves.

All of this is a bit in flux: this is at least blocked until https://github.com/purescript/registry/pull/140 merges.

**Comment by f-f** [comment]
Some more notes on your points:
1. The main thing we need to support for this is the notion of "targets", and indeed there's #681 open to track this change to the config format. I am not sure if we should pivot that issue to track the whole config thing - in any case, I think the new config format is going to look something between the new Manifest type and our current config. In the early days of the Registry RFC the `Package` type (which later became `Manifest`) is basically how I imagine this config will look like. E.g. take a look at the [`Package`](https://github.com/purescript/registry/blob/14a8fb7c6c711ae2bcee54c4753010d57af11721/v1/Package.dhall), [`Target`](https://github.com/purescript/registry/blob/14a8fb7c6c711ae2bcee54c4753010d57af11721/v1/Target.dhall), and [`Index`](https://github.com/purescript/registry/blob/14a8fb7c6c711ae2bcee54c4753010d57af11721/v1/Index.dhall) types.
2. The `dhall` package has [some Template Haskell facilities](https://github.com/dhall-lang/dhall-haskell/blob/master/dhall/src/Dhall/TH.hs) to embed Dhall types into the compilation step so that:
    1. we wouldn't have to type down the Haskell types, and
    2. we'd automatically keep them in sync with upstream.
  
    I.e. we'd basically call `makeHaskellTypes` on [the Operation type](https://raw.githubusercontent.com/purescript/registry/master/v1/Operation.dhall)
3. I would like to avoid directly calling the GitHub API from Spago, because we'd otherwise have to ask the user to configure a GitHub Token. We can avoid that by just printing out a link to create the issue.

To expand a bit more on how the `publish` UX could look like, here's an example:
```
$ spago publish
Generating package manifest from your config...
Running Registry checks..
Checks passed.
Publish your package by opening this ticket: https://github.com/purescript/spago/issues/new?body=etcetc
```

Note: to run the package checks we'd just import the same code that we use for running checks in the Registry CI

**Comment by f-f** [comment]
An update on this now that we have the new codebase in place (see the `spaghetto` folder).
This command would:
- produce a [Manifest](https://github.com/purescript/registry-dev/blob/0bad5ddded9ca7d8f3e461588aa92a05df9125b1/src/Registry/Schema.purs#L21) for the selected package
- run all the Registry checks on the package - as detailed in https://github.com/purescript/registry-dev/issues/523 - so that the publishing operation cannot fail
- produce a link that opens a GitHub issue with the correct payload, or hit the publishing server if that's already up

**Comment by thomashoneyman** [comment]
With regards to the first point, once the new configuration format is stable the registry can support it directly, meaning there is no need to generate a manifest.

**Comment by f-f** [comment]
This is implemented in spaghetto since https://github.com/purescript/spago/commit/b95c82e90cf8a990c622ca09fed09e1b43a6da59.

It will run all the checks and make sure the package is in order for publishing, and the only thing missing is calling out to the registry API, which is blocked on https://github.com/purescript/registry-dev/issues/578
